### PR TITLE
Migrate customer bought and viewed products to grids

### LINKED ISSUE #15444: Reduce client view in Customer Page 1/2
**Is your feature request related to a problem? Please describe.**
Yes, it is. When I look a customer profile, which is a very old and very good customer, my page is really high !!
If we have this important list + some modules, this page become very very slow.
=> I have many customers in this case

**Describe the solution you'd like**
If it's possible to reduce all these informations such as see only 20 last carts, 20 last orders. Same for Products bought by the customer : Top 20 products more bought by the customer.
Every time we want to have more than 20, you have "See more" to display 20 others carts, orders or products.

**Describe alternatives you've considered**
None actually

**Additional context**
Screenshots about only one page/customer : [https://drive.google.com/drive/folders/13TPktWYVdz1Va985wvQH0g0qBHJngt_Z?usp=sharing](https://drive.google.com/drive/folders/13TPktWYVdz1Va985wvQH0g0qBHJngt_Z?usp=sharing)

### PR DESCRIPTION
**Pull request by Hlavtox**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | see below
| Type?             | refacto
| Category?         | BO
| BC breaks?        | partial - CustomerForViewing object now returns empty array for bought and viewed products.
| Deprecations?     | no
| How to test?      | see below
| Fixed ticket?     | Fixes #15444
| Related PRs       | 
| Sponsor company   | 

### Details
- Right now, the bought and viewed product tables on customer details were manually rendered tables. I changed their implementation to grid components like I did with orders and carts in https://github.com/PrestaShop/PrestaShop/pull/31442.
- **This has following advantages:**
  -  Introduces pagination so there is not a kilometer of bought products and the page is actually usable for bigger clients,
  -  Makes the page load faster.
  - Allows the use of standard hooks so people can customize these tables.
  - Bought products - I added a column Order ID, for better clarity.
  - Viewed products - I added a column Cart ID with a link to the cart in backoffice.

### Behavior and wording change
Purchased products now include products from all orders, not just the valid ones. You now see also products from orders that are waiting for payment. Before, they contained only products from valid orders. If the product is bought, it's bought. Follows a simmilar information as in here - https://github.com/PrestaShop/PrestaShop/pull/31440.

Simmilarily, viewed products now contain only products from carts that have not been ordered. ( = not in ps_orders table.) Previously, it contained all products from carts that have not been ordered OR not in a valid order.

The name of `Viewed products` was changed to `Products in carts` to better reflect what it means.

### Screenshot
**Before - chrome did not even want to make me a full height screenshot :D**
https://user-images.githubusercontent.com/6097524/224347878-985f7b3c-5b15-4ff1-8d9f-2f0d7eb8edb9.jpg

**After**
Snímek obrazovky 2023-03-13 142201

### How to test
- Open customer detail page in backoffice and check blocks with bought and viewed products.
- See that everything works the same, but limits the amount of shown products to 20, adds pagination, allows sorting and there is a bit more information. 
- Beware that `Viewed products` block is a bit misleading, what it really means is `Products in unordered carts`.
- Beware that you can see a bit different data with and without this PR. With this PR, you can see more products in `Bought products` and less in `Viewed products`.
- Beware that if customer has no bought products and no products in cart, the block is not shown, like before.

### CODE REVIEW

File: src/Adapter/Customer/QueryHandler/GetCustomerForViewingHandler.php
Comment (matks) [inline src/Adapter/Customer/QueryHandler/GetCustomerForViewingHandler.php:103]: Indeed this is a BC Break

File: src/Core/Grid/Data/Factory/CustomerBoughtProductGridDataFactoryDecorator.php
Comment (matks) [inline src/Core/Grid/Data/Factory/CustomerBoughtProductGridDataFactoryDecorator.php]: Maybe this could look better with sprintf? 😉

File: src/Core/Grid/Data/Factory/CustomerViewedProductGridDataFactoryDecorator.php
Comment (matks) [inline src/Core/Grid/Data/Factory/CustomerViewedProductGridDataFactoryDecorator.php]: I feel we're doing something wrong: you are injecting into a property called product_name some HTML code.

I feel like we are mixing pears and apples.

Maybe we should rather create a new type of column here https://github.com/PrestaShop/PrestaShop/tree/develop/src/Core/Grid/Column/Type/Common ?

Comment (matks) [reply src/Core/Grid/Data/Factory/CustomerViewedProductGridDataFactoryDecorator.php]: ping @zuk3975 what do you think?

Comment (Hlavtox) [reply src/Core/Grid/Data/Factory/CustomerViewedProductGridDataFactoryDecorator.php]: @matks I am using HtmlColumn for that reason ;-)

File: src/Adapter/Customer/QueryHandler/GetCustomerForViewingHandler.php
Comment (Hlavtox) [reply src/Adapter/Customer/QueryHandler/GetCustomerForViewingHandler.php:103]: Yes, I will specify it in the PR description

File: src/Core/Grid/Data/Factory/CustomerViewedProductGridDataFactoryDecorator.php
Comment (Hlavtox) [reply src/Core/Grid/Data/Factory/CustomerViewedProductGridDataFactoryDecorator.php]: Oooh there is a LinkColumn and it supports unstyled links like this, I will try it: 🚀

File: src/Core/Grid/Data/Factory/CustomerBoughtProductGridDataFactoryDecorator.php
Comment (Hlavtox) [reply src/Core/Grid/Data/Factory/CustomerBoughtProductGridDataFactoryDecorator.php]: Resolved by using `LinkColumn`.

File: src/Adapter/Customer/QueryHandler/GetCustomerForViewingHandler.php
Comment (Hlavtox) [reply src/Adapter/Customer/QueryHandler/GetCustomerForViewingHandler.php:103]: I reverted this BC break and kept the `link` for the next major.

File: src/Core/Grid/Data/Factory/CustomerViewedProductGridDataFactoryDecorator.php
Comment (Hlavtox) [reply src/Core/Grid/Data/Factory/CustomerViewedProductGridDataFactoryDecorator.php]: Done, used `LinkColumn`.

File: src/PrestaShopBundle/Resources/config/services/core/grid/doctrine_query_builder.yml
Comment (FabienPapet) [inline src/PrestaShopBundle/Resources/config/services/core/grid/doctrine_query_builder.yml:124]: For all services
```suggestion
    public: false
```

File: src/Adapter/Customer/QueryHandler/GetCustomerForViewingHandler.php
Comment (FabienPapet) [reply src/Adapter/Customer/QueryHandler/GetCustomerForViewingHandler.php:103]: This is not a BC break as the query handler is instanciated by DI

File: src/PrestaShopBundle/Resources/config/services/core/grid/doctrine_query_builder.yml
Comment (kpodemski) [reply src/PrestaShopBundle/Resources/config/services/core/grid/doctrine_query_builder.yml:124]: @FabienPapet is it possible to decorate/override a service marked as `public: false`?

Comment (FabienPapet) [reply src/PrestaShopBundle/Resources/config/services/core/grid/doctrine_query_builder.yml:124]: yes, just like any other service

**Comment by MatShir** [comment]
Nice revamp of the page 🍬 !
We might need to change the wording of the blocks IMO. cc @l-delin 
I suggested 'Viewed products' to 'Abandoned products'.
I suggested 'Purchased product' to 'Ordered products'.
wdyt?

**Comment by Hlavtox** [comment]
`Abandoned products` is not correct, he could have added it to cart today. ;-)

**Comment by l-delin** [comment]
@MatShir I like 'Ordered products', it makes sense since there's the **Orders** page. 👍

**Comment by l-delin** [comment]
@MatShir Let's stick with `Purchased products`, it's a string that already exists in multiple parts of PrestaShop

image 

It's better to keep it for uniformity.

**Comment by MatShir** [comment]
Yes but here IMO the context has evolved, and people can have different definition of purchased.

**Comment by Hlavtox** [comment]
@MatShir @l-delin `Purchased products` is fine for me, I need the second wording validated. 😛

**Comment by MatShir** [comment]
True ! 

> Abandoned products is not correct, he could have added it to cart today. ;-)

Product linked to a cart :P (ugly I know)

**Comment by Hlavtox** [comment]
I suggest `Products in carts`. Best reflects what it means, short.

**Comment by MatShir** [comment]
I let @l-delin take the final decision, I was just giving suggestion :)

**Comment by l-delin** [comment]
`Products in carts` is good! Conveys all you need to know. It's a great suggestion

**Comment by Hlavtox** [comment]
https://www.youtube.com/watch?v=Jo6fKboqfMs

**Comment by MatShir** [comment]
🤣

File: src/Core/Search/Filters/CustomerBoughtProductFilters.php
Comment (zuk3975) [inline src/Core/Search/Filters/CustomerBoughtProductFilters.php:48]: is this custom requirment having this limit?
If not, you can use default limit constant from parent Filters class

File: src/Core/Search/Filters/CustomerViewedProductFilters.php
Comment (zuk3975) [inline src/Core/Search/Filters/CustomerViewedProductFilters.php:48]: same here

**Review by zuk3975 — DISMISSED** [review]

File: src/Core/Search/Filters/CustomerBoughtProductFilters.php
Comment (Hlavtox) [reply src/Core/Search/Filters/CustomerBoughtProductFilters.php:48]: @zuk3975 We opted to limit this to 20 by default, so it's not that long. 👍

File: src/Core/Search/Filters/CustomerViewedProductFilters.php
Comment (Hlavtox) [reply src/Core/Search/Filters/CustomerViewedProductFilters.php:48]: @zuk3975 We opted to limit this to 20 by default, so it's not that long. 👍

**Review by kpodemski — DISMISSED** [review]

**Review by florine2623 — COMMENTED** [review]
Hello @Hlavtox ,

Thanks for the PR!

Link to automated tests : https://github.com/florine2623/testing_pr/actions/runs/4552205627/jobs/8031582281

I have an issue in the customer page in `Products in carts` block. All of the products are not displayed : :x:
In my Cart as a customer I have 29 products :
Screenshot 2023-03-29 at 15 01 05

In my Customer page in BO, only 14 are displayed :
Screenshot 2023-03-29 at 15 00 40

Same type of issue in the block `Purchased products`. :x:
I had 7 products bought before + I process to checkout with 29 products = I should have 36 products. Only 35 are displayed :
Screenshot 2023-03-29 at 15 03 21

Screenshot 2023-03-29 at 15 03 38

Could you check ?
Thanks!

**Comment by Hlavtox** [comment]
@florine2623 Could you provide me with a complete screenshot of that cart in FO + BO, so I can see what is happening on your side?

**Comment by florine2623** [comment]
When I add product to my cart :

In my BO, block `Products in carts` :
Screenshot 2023-03-29 at 15 56 55

In my Cart - 22 products :
screencapture-pr-develop-8888-en-order-2023-03-29-16_00_39

The numbers don't match ☹️ 

When the checkout process is done, now the number of products in `Purchased products` block is correct.
I didn't encounter my previous bug 😓 but I have an error :
Not sure if it is related to your PR or not.
Screenshot 2023-03-29 at 16 22 07

Let me know if that's clear enough, if not we can discuss it together.
Thanks!

**Comment by Hlavtox** [comment]
@florine2623 Hi, yes you are right, I made a mistake and put a `DISTINCT` into the query, which resulted only in unique product IDs to be considered.

Now both the pagination and product count in "Products in carts" section will work properly I hope.
pr

The error you mentioned was a cache issue, it will disappear if you clear cache.

**Review by zuk3975 — DISMISSED** [review]

**Review by kpodemski — DISMISSED** [review]

**Review by HanaRebaiQA — COMMENTED** [review]
Hello @Hlavtox 

I have checked your PR. Here is my feedback : 

**Issue 1 :** The number of Products in carts is not equal to the exact number on FO

image
image

https://user-images.githubusercontent.com/92912932/229537792-31af36e5-9692-455f-9a8a-c4c5f1798c0d.mp4


**Issue 2 :** The pagination of the `Purchased products` refreshes the whole page and it's displayed from the top of the page instead of only the concerned block Purchased products

It can be an improvement. What do you think? 

https://user-images.githubusercontent.com/92912932/229537898-97bc35df-e429-4c23-a7a0-745b71923ca8.mp4

Another issue but it's not related to your PR #32022 

The automated tests failed on Order and Customer. https://github.com/HanaRebaiQA/testing_pr/actions/runs/4594677085/jobs/8115053714
Compared to nightly https://github.com/PrestaShop/PrestaShop/actions/runs/4591953722

ping @PrestaShop/qa-automation What do you think? It's related to this PR ?
**PS** : I launched the failed jobs many times.

Thank You!

**Comment by Hlavtox** [comment]
Hi @HanaRebaiQA

> Issue 1 : The number of Products in carts is not equal to the exact number on FO

The same behavior without this PR. It displays only unique products. There would be 13 lines of the same product :-)

> Issue 2 : The pagination of the Purchased products refreshes the whole page and it's displayed from the top of the page instead of only the concerned block Purchased products

The same behavior on all other pages in backoffice. It could be improved in the future, but in another PR.

> The automated tests failed on Order and Customer. 
https://github.com/HanaRebaiQA/testing_pr/actions/runs/4594677085/jobs/8115053714

Seems related to this PR, automation team probably needs to update the selectors. 👍

**Comment by HanaRebaiQA** [comment]
Thanks @Hlavtox for your feedback! 
So, we wait for the @PrestaShop/qa-automation feedback before validating the PR.

**Comment by nesrineabdmouleh** [comment]
Hello @HanaRebaiQA
The errors of UI tests are related to changed selectors as @Hlavtox said, I'll fix them in the same branch :)
Thank you!

**Comment by HanaRebaiQA** [comment]
Thanks @nesrineabdmouleh for your answer 
So, we can validate this PR ?

**Comment by nesrineabdmouleh** [comment]
Hello @Hlavtox, @HanaRebaiQA 

I fixed UI tests => here is the run https://github.com/nesrineabdmouleh/testing_pr/actions/runs/4606591347
If all is green you can validate the PR ;)

**Review by boubkerbribri — APPROVED** [review]
LGTM UI Tests

**Review by zuk3975 — APPROVED** [review]

**Comment by Hlavtox** [comment]
Thanks everyone!

**Pull request merged**
